torch/distributed/fsdp/_wrap_utils.py

```diff
@@ -119,3 +127,138 @@ def _warn_on_overridden_mixed_precision(
         "These modules will be wrapped as separate FSDP instacnes with mixed "
         "precision disabled."
     )
+
+
+def _validate_frozen_params(
+    root_module: nn.Module,
+    modules_to_wrap: Set[nn.Module],
+    ignored_params: Set[nn.Parameter],
+    use_orig_params: bool,
+):
+    """
+    This checks that, given ``modules_to_wrap``, each module would manage
+    parameters that are uniformly frozen or non-frozen. This uniformity
+    requirement is strict for ``use_orig_params=False`` (hard error) and highly
+    recommended for ``use_orig_params=True`` (user warning).
+    """
+    topo_sorted_named_modules = _get_topo_sorted_named_modules(root_module)
+    reverse_topo_sorted_named_modules = reversed(topo_sorted_named_modules)
+    visited_modules: Set[nn.Module] = set()
+    for module_name, module in reverse_topo_sorted_named_modules:
+        if module in modules_to_wrap:
+            param_to_fqn = _get_managed_param_to_fqn(
+                module, ignored_params, visited_modules, module_name
+            )
+            frozen_param_fqns: List[str] = []
+            frozen_param_numel = 0
+            nonfrozen_param_fqns: List[str] = []
+            nonfrozen_param_numel = 0
+            for param, fqn in param_to_fqn.items():
+                if param.requires_grad:
+                    nonfrozen_param_fqns.append(fqn)
+                    nonfrozen_param_numel += param.numel()
+                else:
+                    frozen_param_fqns.append(fqn)
+                    frozen_param_numel += param.numel()
+            if len(frozen_param_fqns) > 0 and len(nonfrozen_param_fqns) > 0:
+                msg = f"{module_name} has both parameters with requires_grad=True and False."
+                if use_orig_params:
+                    total_param_numel = frozen_param_numel + nonfrozen_param_numel
+                    msg += (
+                        " We do not recommend wrapping such modules since "
+                        "the gradient memory usage will be higher than expected "
+                        f"({total_param_numel} numel instead of {nonfrozen_param_numel} numel "
+                        "before sharding via reduce-scatter). "
+                    )
+                else:
+                    msg += " FSDP does not support wrapping such modules when use_orig_params=False. "
+                msg += "If possible, wrap the frozen parameters with FSDP separately.\n"
```

Comment:
In general, this requires splitting your module into multiple modules, some of which have frozen params and some of which don't, right? Is it always safe to take a module M and create two submodules beneath it M1 and M2, one of which has frozen params and the other which doesn't (even if I don't call M1.forward/M2.forward?)

Reply:
> In general, this requires splitting your module into multiple modules, some of which have frozen params and some of which don't, right?

Yes.

> Is it always safe to take a module M and create two submodules beneath it M1 and M2, one of which has frozen params and the other which doesn't (even if I don't call M1.forward/M2.forward?)

Unfortunately, it is important to still call M1.forward/M2.forward in order to run pre/post-forward logic, which would notably include the corresponding all-gathers in the pre-forward. In other words, if we just organize the parameters into modules but do not change the parent's forward to actually call those modules' forward, then they will not have their parameters all-gathered.